### DISCUSSION TITLE
Add a new sensor device class: duration

### DISCUSSION THREAD

**Discussion by bdraco** [home-assistant/architecture#762] [Entity Models] [answered]
The goal is to avoid integrations adding display layer formatted data in the state machine.

The example use case is `history_stats`.  We currently store the duration as the state and as an attribute.  The attribute was only added to make the time human readable on the frontend. 
<img width="386" alt="Screen Shot 2022-04-21 at 12 07 06" src="https://user-images.githubusercontent.com/663432/164559635-69eaeb68-d761-43c6-86c3-9603da0c0dfc.png"><img width="390" alt="Screen Shot 2022-04-21 at 12 41 06" src="https://user-images.githubusercontent.com/663432/164563070-906f1db6-9062-455f-a2d1-a3320a64c229.png">

- We expect to be able to use this for timers as well.

- This would only be used for fixed/stopped durations between two points in time where a `timestamp` class could not be used.

- The frontend will be able to format the duration into a human readable time using something similar to https://github.com/home-assistant/frontend/blob/dev/src/common/datetime/seconds_to_duration.ts

PRs:
https://github.com/home-assistant/core/pull/70543

**Comment by balloob** [comment]
Thinking about it now, I think that we should allow a duration to work _like timestamp_ without actually setting the `timestamp` class. Probably should be modeled like our media player `position` property. For media player we have `position`, `position_updated_at` and the media player state (playing, paused).

The computed position of a media player is set to:

- If paused: position
- If playing: position + (now - position_updated_at)

We could adopt something similar for a duration.

**Reply by bdraco** [reply]
An alternative would be no backend change, and if `unit_of_measurement` is 

```
# Time units
TIME_MICROSECONDS: Final = "μs"
TIME_MILLISECONDS: Final = "ms"
TIME_SECONDS: Final = "s"
TIME_MINUTES: Final = "min"
TIME_HOURS: Final = "h"
```

Then we display it on the frontend with with https://github.com/home-assistant/frontend/blob/dev/src/common/datetime/seconds_to_duration.ts ?  (we would need to teach it about `μs` and `ms`)

**Reply by bdraco** [reply]
~~That gets a bit weird though since the unit of measurement would be value would display as `02:59:02` and the uom as `h`~~

Actually looking at https://github.com/home-assistant/frontend/blob/dev/src/common/entity/compute_state_display.ts#L112 the display state is a single field so I think it would be ok

**Reply by bdraco** [reply]
> The computed position of a media player is set to:
> 
> * If paused: position
> * If playing: position + (now - position_updated_at)
> 
> We could adopt something similar for a duration.

At least for `history_stats`, we can't compute the state on the frontend since it is the time the state was `X` between `time1` and `time2`.

**Reply by elupus** [reply]
Couldn't it be split in two pieces? Non active duration, currently active duration. Non active being a non changing number, active being a timestamp when activated.

**Reply by bdraco** [reply]
> Couldn't it be split in two pieces? Non active duration, currently active duration. Non active being a non changing number, active being a timestamp when activated.

Can you expand on that a bit?  I'm not following how that would work with the example pool pump sensor above.

**Reply by elupus** [reply]
So history stats accumulates the total time it has been on historically. Excluding if it's on right now. Exposes that as state. We add an attribute (similar to last updated in the media player case) named activated_at.

Frontend can then calculate the total on time as `historic_on + iif(activated_at, now - activated_at, 0)`

**Reply by bdraco** [reply]
Right now the state of `history_stats` is the total amount of time the device is in the matching states (`on`, `blue`, `off`, `red`, etc) updated each time the sensor updates. It would be quite a breaking change if we changed how that worked as these are used as the basis for automations.

I used to have something like the below psudeo code before great Texas freeze power meltdown (when you could actually get a reasonable variable rate): 

if power_is_cheap and pool_pump_hours_run < 10:
  ...turn on pump

if pool_pump_hours_run > 14:
  ...turn off pump

**Reply by elupus** [reply]
Indeed. Not sure how to overcome that and keep state updates low.

**Reply by balloob** [reply]
I agree things do get complicated if you want to distinguish between active/passive durations.

- Duration trigger testing when >10h would have to calculate a point in time for an active duration to pass that.
- Duration condition would probably be a template helper `{{ test_duration("sensor.pool_pump_active", ">", "10h") }}`

**Comment by davet2001** [comment]
Could it be used for cumulative durations made of many short durations? Eg total time fan was on in a week?

**Reply by bdraco** [reply]
> total time fan was on in a week?

That's one of the use cases for a `history_stats` sensor

**Comment by frenck** [comment, marked as answer]
Implemented.
